source/common/event/guarddog_impl.h

```diff
@@ -0,0 +1,87 @@
+#pragma once
+
+#include "envoy/event/guarddog.h"
+#include "envoy/event/watchdog.h"
+#include "envoy/server/configuration.h"
+#include "envoy/stats/stats.h"
+
+#include "common/common/logger.h"
+#include "common/common/thread.h"
+#include "common/event/libevent.h"
+
+namespace Event {
+
+/** This feature performs deadlock detection stats collection&enforcement.
+ *
+ * It launches a thread that scans at an interval the minimum of the configured
+ * intervals.  If it finds starved threads or suspected deadlocks it will take
+ * the appropriate action depending on the config parameters described below.
+ *
+ * Thread lifetime is tied to GuardDog object lifetime (RAII style).
+ */
+class GuardDogImpl : public GuardDog {
+public:
+  /**
+   * @param stats_store Statistics store to write watchdog_miss and
+   * watchdog_mega_miss events into.
+   * @param config Configuration object.
+   *
+   * Configuration:
+   *   1) Miss timeout: Record a "miss" counter increment after this long has
+   *   passed.  Old default: 100ms.
```

Comment:
This was mostly a note for myself, I'll clean it up.  The old default is a reference to the hardcoded values from the original non-threaded watchdog stats.